microprofile/WFLY-10709-MicroProfile-Tracing-Extension.adoc

```diff
@@ -0,0 +1,79 @@
+= [WFLY-10709] Support Eclipse MicroProfile OpenTracing
+:author:            Juraci Paixão Kröhling
+:email:             jpkroehling@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:keywords:          observability,microprofile,tracing,openshift
+
+== Overview
+
+In order to improve Observability of applications deployed in WildFly it would be beneficial to provide
+transparent tracing of Java EE components via automatic initialization of a tracer (Jaeger) plus a set
+of instrumentation libraries for components commonly used in Java EE applications, such as Servlets, JAX-RS,
+CDI, EJB, ...
+
+This will be implemented by supporting Eclipse MicroProfile OpenTracing 1.1 for applications deployed in WildFly.
+The integration uses the smallrye-opentracing component to provide the MicroProfile OpenTracing implementation.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-10709[WFLY-10709]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1027[EAP7-1027]
+* https://issues.jboss.org/browse/CLOUD-2728[CLOUD-2728] - Expose EAP "Tracing" information so that it can be consumed by OpenShift Console
+* https://issues.jboss.org/browse/TRACING-85[TRACING-85] - Create jboss-modules for OpenTracing and Jaeger
+* https://issues.jboss.org/browse/TRACING-86[TRACING-86] - OpenTracing integration for EJB in Wildfly
+* https://issues.jboss.org/browse/TRACING-87[TRACING-87] - OpenTracing integration for JAX-RS in Wildfly
+* https://issues.jboss.org/browse/TRACING-88[TRACING-88] - OpenTracing integration for JMS in Wildfly
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:tremes@redhat.com[Tomas Remes]
+* mailto:jstourac@redhat.com[Jan Stourac]
+
+=== Affected Projects or Components
+
+The implementation of Eclipse MicroProfile OpenTracing is provided by https://github.com/smallrye/smallrye-opentracing[smallrye-opentracing].
+
+=== Other Interested Projects
+
+* A https://github.com/jpkrohling/java-wildfly-tracing[prototype of a WildFly extension] for OpenTracing was developed while exploring a solution for TRACING-85/.../TRACING-88.
+
+== Requirements
+
+* Support Eclipse MicroProfile OpenTracing 1.1 for applications deployed in WildFly.
+* Support a concrete tracer, such as Jaeger's
```

Comment:
I think it's required that the concrete tracer must be configurable in a way consistent with general WildFly configuration standards. Things like proper use of paths, [outbound]-socket-bindings.

Details on what that config would look like would be good. A lot of times a draft those details surfaces requirements.

Reply:
Added some details. I'm probably a bit outdated regarding WF configuration, but would would the outbound socket-binding you mentioned be something relevant for `JAEGER_AGENT_HOST`, `JAEGER_AGENT_PORT` and `JAEGER_COLLECTOR_ENDPOINT` ?

Reply:
Sounds like it, yes. A remote-outbound-socket-binding is used for declaring the host and port of a remote service to which an in-server client will connect.

This is something we should chat about as I can see how this might be a pain in a cloud use case. On OpenShift we provide a standard standalone-openshift.xml and then based on env var the user provides (e.g. when create their app from an application template) scripts modify the config.  The more complex the script modifications, the harder it is.  We'll have to work out how the scripts will work regardless.

Reply:
Currently, the Jaeger Java Client can be almost completely configured via env vars and system properties. The cases where it cannot (like, when there's a custom `RemoteReporter` or `Sender`) would also (probably) not be covered by configuring it via WildFly's standard mechanism.

Additionally, this configuration is Jaeger-specific and I remember that it was said during the kick-off meeting that it would be a good idea to think twice about the configuration for this first version, as whatever we decide now is harder to change later, especially considering that we might support other tracers in the future (with a different set of configuration options).

That all said, could we then just document that the tracer can be configured by setting env vars/system properties, either via:

* real env vars (for Docker, Kubernetes, ...)
* `-u JAEGER_SERVICE_NAME=service-name`
* `-P file:///path/to/configuration.properties`
* `<server><system-properties><property name="JAEGER_SERVICE_NAME" value="service-name"/></system-properties>...</server>`

We could probably also have a `<properties/>` as a child of the subsystem, like `management/security-realms/security-realm/authentication/properties`, but I don't see how it would bring any benefits over the methods above.

Reply:
@jpkrohling Sorry, I didn't see this response before.

Configuring WildFly via env vars / system properties is quite contrary to how things are supposed to be done.

Re: thinking hard about the configuration because it's hard to change later, that applies to configuring via env vars and properties as well. Once this is in, the specific env vars that are used now have to work forever. Of course we could say things in standalone.xml take precedence over the env vars. But the default behavior with no config beyond having the subsystem would have to remain as we do it now.

In the future I don't much like a <properties/> child of the subsystem. As a way of configuring a plugin provider, sure. But our standard provider should be configurable via fully specified attributes.

@jmesnil FYI.

My inclination is this should be a Tech Preview feature until the config model is clarified. But I think providing it configured via env vars is ok. The TP basically gives us freedom to change the config model in the future if we decide we want to

Reply:
> Once this is in, the specific env vars that are used now have to work forever

This is already true for Jaeger and how it's been documented to work on other stacks.

> The TP basically gives us freedom to change the config model in the future if we decide we want to

+1

Reply:
@bstansberry Even though we mark this as a TP, we need to consider adding proper means of configuration for WildFly. While for CD, this is no problem, for EAP zip stream, this has to be done properly. It is not realistic to include new subsystem out of blue in CP patches in proper stream.

Also, even though we mark this as a TP, this feature will be adopted by all layered products based on EAP (this will be their default go-to solution for Kiali), thus any changes in its configuration model will have to be done carefully and with backward compatibility in mind. So we should try to make this as much final as possible.

Configuration of tracing agent via WildFly management is therefore necessary. We also need to consider that on OpenShift, the management will not be too useful, so default configuration of tracing agent should utilise expression, like how it is currently done e.g. for:
```
/socket-binding-group=standard-sockets/socket-binding=http[port]

ls /socket-binding-group=standard-sockets/socket-binding=http:read-resource-description
bound=true                    bound-port=8080               fixed-port=false              multicast-address=undefined   name=http                     
bound-address=127.0.0.1       client-mappings=undefined     interface=undefined           multicast-port=undefined      port=${jboss.http.port:8080}
```
Also, looking at the list of [JAEGER properties](https://github.com/jaegertracing/jaeger-client-java/tree/master/jaeger-core#configuration-via-environment), when we implement those to be configurable via management model, we need to use existing infrastructure in WildFly - namely use socket bindings for information about connections, Elytron to hold security, etc.

Reply:
> we need to consider adding proper means of configuration for WildFly

Other than tracer-specific configuration, I'm not sure what else there is to configure. Perhaps a switch to turn tracing on/off? This would probably require a re-deploy of the affected applications, as filters/interceptors would need to be removed.

> any changes in its configuration model will have to be done carefully and with backward compatibility in mind

That's exactly why I'm a bit reluctant in adding non-mandatory configuration options now. In general, it's easier to keep backwards compatibility when we add options than when we move/remove them. I believe this is true for WildFly as well.

> Also, looking at the list of JAEGER properties, when we implement those to be configurable via management model, we need to use existing infrastructure in WildFly - namely use socket bindings for information about connections, Elytron to hold security, etc.

None of those properties are required, except for the service name, which we can derive from the deployment unit's name.

Not sure how Elytron fits here, would you mind expanding on that? 

About the socket bindings: those are Jaeger-specific settings and we discussed in the past that we *could* envision users wanting to make use of their own tracers (from a third-party vendor or home-made), either as part of their deployments or as a custom JBoss Module (like a JDBC driver), which is why I wanted to avoid having Jaeger-specific information in the configuration model. If we do want to have Jaeger-specific configuration under WildFly's management model, I'm OK with that. I would just need a suggestion on how to best do that in a way that we can change it in the future.

Reply:
Yes, on/off switch is a good thing to have. It is easier to control than adding/removing the whole extension IMHO - at least from the UX point of view.

Our point is to have a possibility to configure tracing agent via management model. I believe there is some common set of attributes that can be used to configure various tracing agents? E.g. user/password (that was the reason why I mentioned elytron integration in previous comment). Such set should be determined and added into the `microprofile-opentracing-smallrye` subsystem. Maybe @bstansberry can provide any opinion?

Regarding Elytron - please see `authentication-confguration` and `authentication-context` server configuration elements. These are used when server is authenticating against remote security provider and it would be useful if tracing agents were able to use the same mechanisms as the rest of endpoints on the server. cc @darranl

Reply:
> Yes, on/off switch is a good thing to have. It is easier to control than adding/removing the whole extension IMHO - at least from the UX point of view.

Would you say this is required for the Tech Preview? My understand is that the final PR has to be ready by this Friday (two days from now), and the proposal hasn't been merged yet.

If this is required, I'll start working on it right away, but I'd really, really appreciate to have a list of things that are required to be ready by this Friday.

> E.g. user/password

There's no auth/z in Jaeger. So far, the current approach in protecting Jaeger's UI is to place an OAuth proxy in front of it, which is why we have a token and username/password env vars, but I don't think this setup will be supported in the product.